apis/v1alpha1/backendpolicy_types.go

```diff
@@ -0,0 +1,173 @@
+// +kubebuilder:subresource:status
+
+// BackendPolicy defines policies associated with backends. For the purpose of
+// this API, a backend is defined as any resource that a route can forward
+// traffic to. A common example of a backend is a Service. Configuration that is
+// implementation specific may be represented with similar implementation
+// specific custom resources.
+type BackendPolicy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   BackendPolicySpec   `json:"spec,omitempty"`
+	Status BackendPolicyStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// BackendPolicyList contains a list of BackendPolicy
+type BackendPolicyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []BackendPolicy `json:"items"`
+}
+
+// BackendPolicySpec defines desired policy for a backend.
+type BackendPolicySpec struct {
+	// BackendRefs define which backends this policy should be applied to. If
+	// more than one BackendPolicy targets the same backend, precedence will be
+	// given to the oldest BackendPolicy.
+	//
+	// Support: Core
+	//
+	// +optional
+	BackendRefs []BackendRef `json:"backendRefs"`
+
+	// TLS is the TLS configuration for these backends.
+	//
+	// Support: Extended
+	//
+	// +optional
+	TLS *BackendTLSConfig `json:"tls,omitempty"`
+}
+
+// BackendRef identifies an API object within a known namespace that defaults
+// group to core and resource to services if unspecified.
+type BackendRef struct {
+	// Group is the group of the referent.  Omitting the value or specifying
+	// the empty string indicates the core API group.  For example, use the
+	// following to specify a service:
+	//
+	// fooRef:
+	//   resource: services
+	//   name: myservice
+	//
+	// Otherwise, if the core API group is not desired, specify the desired
+	// group:
+	//
+	// fooRef:
+	//   group: acme.io
+	//   resource: foos
+	//   name: myfoo
+	//
+	// +optional
+	// +kubebuilder:default=core
+	Group string `json:"group"`
+
+	// Resource is the API resource name of the referent. Omitting the value
+	// or specifying the empty string indicates the services resource. For example,
+	// use the following to specify a services resource:
+	//
+	// fooRef:
+	//   name: myservice
+	//
+	// Otherwise, if the services resource is not desired, specify the desired
+	// group:
+	//
+	// fooRef:
+	//   group: acme.io
+	//   resource: foos
+	//   name: myfoo
+	//
+	// +optional
+	// +kubebuilder:default=services
+	Resource string `json:"resource"`
+
+	// Name is the name of the referent.
+	//
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+
+	// Port is the port of the referent. If unspecified, this policy applies to
+	// all ports on the backend.
+	// +optional
+	Port *int32 `json:"port,omitempty"`
+}
+
+// BackendTLSConfig describes TLS configuration for a backend.
+type BackendTLSConfig struct {
```

Comment:
@robscott what is the reason for being able to specify an empty `BackendTLSConfig`? If no reason exists, should the client cert ref be required?

Reply:
That's a good question. I think it made sense as an optional field on GatewayTLSConfig but I can't think of any reason it needs to be optional here, will update.

Reply:
Updated to make this required, thanks!